trl/data_utils.py

```diff
@@ -644,23 +644,71 @@ def search(self, val):
 
 def _pack_bfd(examples: pa.Table, seq_length: int) -> pa.Table:
     """Pack sequences in a pyarrow Table using Best Fit Decreasing strategy."""
+    # Identify the list column and prepare all columns
     columns = []
     list_column_idx = None
     for idx, column in enumerate(examples.columns):
-        if pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type):
-            column = pc.list_slice(column, 0, seq_length)
-            if list_column_idx is None:
-                list_column_idx = idx
+        if isinstance(column, pa.ChunkedArray):
+            column = column.combine_chunks()
+        if not (pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type)):
+            raise TypeError("pack_dataset(bfd) requires all columns to be list-like.")
+        if list_column_idx is None:
+            list_column_idx = idx
         columns.append(column)
-    examples = pa.Table.from_arrays(columns, names=examples.column_names)
 
-    ids = np.arange(len(examples))
     assert list_column_idx is not None
+    list_column = columns[list_column_idx]
+    offsets = np.asarray(list_column.offsets)
+    values = list_column.values
+
+    # Split every list row into fragments of length <= seq_length (so long rows become multiple samples).
+    frag_lengths: list[int] = []
+    frag_info: list[tuple[int, int, int]] = []  # (row_idx, split_start, frag_len)
+    expanded_indices: list[int] = []
+    for row_idx, (row_start, row_end) in enumerate(zip(offsets[:-1], offsets[1:], strict=False)):
+        length = row_end - row_start
+        for split_start in range(0, length, seq_length):
+            frag_len = min(seq_length, length - split_start)
+            frag_lengths.append(frag_len)
+            frag_info.append((row_idx, split_start, frag_len))
+            expanded_indices.append(row_idx)
+
+    # Rebuild list columns with fragments and duplicate non-list columns accordingly.
```

Comment:
Maybe we can remove the no longer applicable part of the comment:

```suggestion
    # Rebuild list columns with fragments
```

Reply:
Right, I can't remember where this comment comes from.